File: step_08_bit_manipulation/8.3_advanced_maths/P004_find_prime_factorization_using_sieve/Solution.java
/**
 * Problem: Find Prime Factorization using Sieve
 * Difficulty: MEDIUM | XP: 25
 *
 * @author DSA_Nova
 */
import java.util.*;

public class Solution {

    // ============================================================
    // APPROACH 1: BRUTE FORCE - Trial division
    // Time: O(sqrt(n))  |  Space: O(log n)
    // ============================================================
    public static List<int[]> bruteForce(int n) {
        // Returns list of [prime, exponent] pairs
        List<int[]> factors = new ArrayList<>();
        for (int i = 2; i * i <= n; i++) {
            int count = 0;
            while (n % i == 0) {
                count++;
                n /= i;
            }
            if (count > 0) factors.add(new int[]{i, count});
        }
        if (n > 1) factors.add(new int[]{n, 1});
        return factors;
    }

    // ============================================================
    // APPROACH 2: OPTIMAL - SPF (Smallest Prime Factor) sieve + factorize
    // Time: O(n log log n) precompute, O(log n) per query  |  Space: O(n)
    // ============================================================
    static int[] spf; // smallest prime factor

    public static void buildSPF(int limit) {
        spf = new int[limit + 1];
        for (int i = 0; i <= limit; i++) spf[i] = i;
        for (int i = 2; (long) i * i <= limit; i++) {
            if (spf[i] == i) { // i is prime
                for (int j = i * i; j <= limit; j += i) {
                    if (spf[j] == j) spf[j] = i;
                }
            }
        }
    }

    public static List<int[]> optimal(int n) {
        if (spf == null || spf.length <= n) buildSPF(n);
        List<int[]> factors = new ArrayList<>();
        while (n > 1) {
            int p = spf[n];
            int count = 0;
            while (n % p == 0) {
                count++;
                n /= p;
            }
            factors.add(new int[]{p, count});
        }
        return factors;
    }

    // ============================================================
    // APPROACH 3: BEST - SPF sieve with log-based factorization
    // Time: O(n) linear sieve precompute, O(log n) per query  |  Space: O(n)
    // ============================================================
    static int[] linearSpf;

    public static void buildLinearSieve(int limit) {
        linearSpf = new int[limit + 1];
        List<Integer> primes = new ArrayList<>();
        for (int i = 2; i <= limit; i++) {
            if (linearSpf[i] == 0) {
                linearSpf[i] = i;
                primes.add(i);
            }
            for (int p : primes) {
                if (p > linearSpf[i] || (long) i * p > limit) break;
                linearSpf[i * p] = p;
            }
        }
    }

    public static List<int[]> best(int n) {
        if (linearSpf == null || linearSpf.length <= n) buildLinearSieve(n);
        List<int[]> factors = new ArrayList<>();
        while (n > 1) {
            int p = linearSpf[n];
            int count = 0;
            while (n % p == 0) {
                count++;
                n /= p;
            }
            factors.add(new int[]{p, count});
        }
        return factors;
    }

    private static String formatFactors(List<int[]> factors) {
        StringBuilder sb = new StringBuilder();
        for (int[] f : factors) {
            if (sb.length() > 0) sb.append(" * ");
            sb.append(f[0]);
            if (f[1] > 1) sb.append("^").append(f[1]);
        }
        return sb.toString();
    }

    public static void main(String[] args) {
        System.out.println("=== Find Prime Factorization using Sieve ===");
        int n = 360;
        System.out.println("Brute Force: " + n + " = " + formatFactors(bruteForce(n)));
        System.out.println("Optimal:     " + n + " = " + formatFactors(optimal(n)));
        System.out.println("Best:        " + n + " = " + formatFactors(best(n)));
    }
}

File: step_08_bit_manipulation/8.3_advanced_maths/P004_find_prime_factorization_using_sieve/solution.py
"""
Problem: Find Prime Factorization using Sieve
Difficulty: MEDIUM | XP: 25
"""
from typing import List, Dict


# ============================================================
# APPROACH 1: BRUTE FORCE — Trial division per query
# Time: O(sqrt(n)) per query  |  Space: O(log n) result
# ============================================================
def brute_force(n: int) -> List[int]:
    """
    Divide n by 2, then odd numbers up to sqrt(n).
    Returns sorted list of prime factors (with repetition).
    e.g. brute_force(12) -> [2, 2, 3]
    """
    factors: List[int] = []
    # Divide out 2s
    while n % 2 == 0:
        factors.append(2)
        n //= 2
    # Odd divisors
    d = 3
    while d * d <= n:
        while n % d == 0:
            factors.append(d)
            n //= d
        d += 2
    if n > 1:
        factors.append(n)
    return factors


# ============================================================
# APPROACH 2: OPTIMAL — Smallest Prime Factor (SPF) sieve
# Preprocessing: O(n log log n)  |  Per query: O(log n)
# Build SPF[i] = smallest prime that divides i, then factor
# any m <= n in O(log m) by repeated lookup.
# ============================================================
def build_spf(limit: int) -> List[int]:
    """Build Smallest Prime Factor sieve up to `limit`."""
    spf = list(range(limit + 1))   # spf[i] = i initially
    p = 2
    while p * p <= limit:
        if spf[p] == p:             # p is prime
            for multiple in range(p * p, limit + 1, p):
                if spf[multiple] == multiple:   # not yet assigned
                    spf[multiple] = p
        p += 1
    return spf


def optimal(n: int, spf: List[int]) -> List[int]:
    """
    Factorise n using prebuilt SPF table.
    Divide by spf[n] repeatedly until n == 1.
    """
    factors: List[int] = []
    while n > 1:
        factors.append(spf[n])
        n //= spf[n]
    return factors


# ============================================================
# APPROACH 3: BEST — SPF sieve returning factor -> exponent map
# Same preprocessing, result as a clean {prime: exp} dict.
# ============================================================
def best(n: int, spf: List[int]) -> Dict[int, int]:
    """
    Same O(log n) factorisation but returns a dictionary
    {prime: exponent} for convenient downstream use.
    e.g. best(12, spf) -> {2: 2, 3: 1}
    """
    factors: Dict[int, int] = {}
    while n > 1:
        p = spf[n]
        while n % p == 0:
            factors[p] = factors.get(p, 0) + 1
            n //= p
    return factors


if __name__ == "__main__":
    print("=== Find Prime Factorization using Sieve ===")
    LIMIT = 100
    spf_table = build_spf(LIMIT)

    test_cases = [12, 36, 97, 84, 100]
    for num in test_cases:
        bf = brute_force(num)
        opt = optimal(num, spf_table)
        bst = best(num, spf_table)
        print(f"n={num:3d} | Brute: {bf} | Optimal: {opt} | Best: {bst}")
